File: ros_ws/bw_depthai/src/depthai_oak_1_w.cpp
#include "depthai_oak_1_w.h"

DepthAiOak1W::DepthAiOak1W(ros::NodeHandle *nodehandle) : nh(*nodehandle),
                                                          _image_transport(nh)
{
    std::string resolution_name;
    ros::param::param<std::string>("~resolution", resolution_name, "1080_P");
    ros::param::param<std::string>("~camera_name", _camera_name, "oak");
    int queue_size;
    ros::param::param<int>("~queue_size", queue_size, 10);
    ros::param::param<int>("~fps", _fps, 60);
    ros::param::param<double>("~camera_matrix_alpha", _alpha, 0.0);

    if (resolution_name.compare("1080_P") == 0)
    {
        _resolution = dai::ColorCameraProperties::SensorResolution::THE_1080_P;
        _width = 1920;
        _height = 1080;
    }
    else if (resolution_name.compare("4_K") == 0)
    {
        _resolution = dai::ColorCameraProperties::SensorResolution::THE_4_K;
        _width = 3840;
        _height = 2160;
    }
    else if (resolution_name.compare("12_MP") == 0)
    {
        _resolution = dai::ColorCameraProperties::SensorResolution::THE_12_MP;
        _width = 4056;
        _height = 3040;
    }
    else if (resolution_name.compare("1352X1012") == 0)
    {
        _resolution = dai::ColorCameraProperties::SensorResolution::THE_1352X1012;
        _width = 1352;
        _height = 1012;
    }
    else if (resolution_name.compare("2024X1520") == 0)
    {
        _resolution = dai::ColorCameraProperties::SensorResolution::THE_2024X1520;
        _width = 2024;
        _height = 1520;
    }
    else
    {
        ROS_ERROR("Invalid resolution name: %s", resolution_name.c_str());
        ros::shutdown();
    }

    _camera_pub = _image_transport.advertiseCamera(_camera_name + "/image_raw", queue_size);
    _rect_image_pub = _image_transport.advertise(_camera_name + "/image_rect", queue_size);
}

DepthAiOak1W::~DepthAiOak1W()
{
}

ros::Time DepthAiOak1W::getFrameTime(ros::Time rosBaseTime,
                                     std::chrono::time_point<std::chrono::steady_clock> steadyBaseTime,
                                     std::chrono::time_point<std::chrono::steady_clock, std::chrono::steady_clock::duration> currTimePoint)
{
    auto elapsedTime = currTimePoint - steadyBaseTime;
    uint64_t nSec = rosBaseTime.toNSec() + std::chrono::duration_cast<std::chrono::nanoseconds>(elapsedTime).count();
    auto currTime = rosBaseTime;
    auto rosStamp = currTime.fromNSec(nSec);
    return rosStamp;
}

void DepthAiOak1W::updateBaseTime(std::chrono::time_point<std::chrono::steady_clock> steadyBaseTime, ros::Time &rosBaseTime)
{
    ros::Time currentRosTime = ros::Time::now();
    std::chrono::time_point<std::chrono::steady_clock> currentSteadyTime = std::chrono::steady_clock::now();
    // In nanoseconds
    auto expectedOffset = std::chrono::duration_cast<std::chrono::nanoseconds>(currentSteadyTime - steadyBaseTime).count();
    uint64_t previousBaseTimeNs = rosBaseTime.toNSec();
    rosBaseTime = rosBaseTime.fromNSec(currentRosTime.toNSec() - expectedOffset);
}

sensor_msgs::CameraInfo DepthAiOak1W::createCameraInfo(int width, int height, std::vector<std::vector<float>> camIntrinsics, std::vector<float> distCoeffs)
{
    sensor_msgs::CameraInfo cameraData;
    cameraData.width = width;
    cameraData.height = height;
    cameraData.header.frame_id = _camera_name;

    std::vector<double> flatIntrinsics, distCoeffsDouble;

    flatIntrinsics.resize(9);
    for (int i = 0; i < 3; i++)
    {
        std::copy(camIntrinsics[i].begin(), camIntrinsics[i].end(), flatIntrinsics.begin() + 3 * i);
    }

    auto &intrinsics = cameraData.K;
    auto &distortions = cameraData.D;
    auto &projection = cameraData.P;
    auto &rotation = cameraData.R;
    // Set rotation to reasonable default even for non-stereo pairs
    rotation[0] = rotation[4] = rotation[8] = 1;
    for (size_t i = 0; i < 3; i++)
    {
        std::copy(flatIntrinsics.begin() + i * 3, flatIntrinsics.begin() + (i + 1) * 3, projection.begin() + i * 4);
    }
    std::copy(flatIntrinsics.begin(), flatIntrinsics.end(), intrinsics.begin());

    for (size_t i = 0; i < 8; i++)
    {
        double distCoeff = static_cast<double>(distCoeffs[i]);
        distortions.push_back(distCoeff);
        distCoeffsDouble.push_back(distCoeff);
    }
    cameraData.distortion_model = "rational_polynomial";

    return cameraData;
}
void DepthAiOak1W::resizeCameraInfo(sensor_msgs::CameraInfo &cameraInfoMsg, int destinationWidth, int destinationHeight)
{
    float scale_y = (float)destinationHeight / cameraInfoMsg.height;
    float scale_x = (float)destinationWidth / cameraInfoMsg.width;
    cameraInfoMsg.height = destinationHeight;
    cameraInfoMsg.width = destinationWidth;

    cameraInfoMsg.K[0] *= scale_x; // fx
    cameraInfoMsg.K[2] *= scale_x; // cx
    cameraInfoMsg.K[4] *= scale_y; // fy
    cameraInfoMsg.K[5] *= scale_y; // cy

    cameraInfoMsg.P[0] *= scale_x; // fx
    cameraInfoMsg.P[2] *= scale_x; // cx
    cameraInfoMsg.P[3] *= scale_x; // T
    cameraInfoMsg.P[5] *= scale_y; // fy
    cameraInfoMsg.P[6] *= scale_y; // cy

    cameraInfoMsg.roi.x_offset = (unsigned int)(cameraInfoMsg.roi.x_offset * scale_x);
    cameraInfoMsg.roi.y_offset = (unsigned int)(cameraInfoMsg.roi.y_offset * scale_y);
    cameraInfoMsg.roi.width = (unsigned int)(cameraInfoMsg.roi.width * scale_x);
    cameraInfoMsg.roi.height = (unsigned int)(cameraInfoMsg.roi.height * scale_y);
}

int DepthAiOak1W::run()
{
    // Create pipeline
    dai::Pipeline pipeline;

    // Define source and output
    auto camRgb = pipeline.create<dai::node::ColorCamera>();
    auto xoutVideo = pipeline.create<dai::node::XLinkOut>();

    xoutVideo->setStreamName("video");

    // Properties
    dai::CameraBoardSocket cameraId = dai::CameraBoardSocket::CAM_A;
    camRgb->setBoardSocket(cameraId);
    camRgb->setResolution(_resolution);
    camRgb->setFps(_fps);

    xoutVideo->input.setBlocking(false);
    xoutVideo->input.setQueueSize(1);

    // Linking
    camRgb->video.link(xoutVideo->input);

    // Connect to device and start pipeline
    dai::Device device(pipeline);

    auto video = device.getOutputQueue("video");

    // Get camera intrinsics
    auto calibHandler = device.readCalibration();

    int width, height;
    std::tie(std::ignore, width, height) = calibHandler.getDefaultIntrinsics(cameraId);

    std::vector<std::vector<float>> camIntrinsics;
    std::vector<float> distCoeffs;
    camIntrinsics = calibHandler.getCameraIntrinsics(cameraId, width, height, dai::Point2f(), dai::Point2f());
    distCoeffs = calibHandler.getDistortionCoefficients(cameraId);
    sensor_msgs::CameraInfo cameraData = createCameraInfo(width, height, camIntrinsics, distCoeffs);
    resizeCameraInfo(cameraData, _width, _height);

    cv::Mat camMatrix(3, 3, CV_32F);
    for (int i = 0; i < 3; i++)
    {
        for (int j = 0; j < 3; j++)
        {
            camMatrix.at<float>(i, j) = cameraData.K[i * 3 + j];
        }
    }

    cv::Size size(_width, _height);
    cv::Mat newMatrix = cv::getOptimalNewCameraMatrix(
        camMatrix,
        cv::Mat(distCoeffs),
        size,
        _alpha,
        size);
    newMatrix.at<double>(2, 2) = 1.0;

    cv::Mat mapX, mapY;
    cv::initUndistortRectifyMap(
        camMatrix,
        cv::Mat(distCoeffs),
        cv::Mat(),
        newMatrix,
        size,
        CV_32FC1,
        mapX,
        mapY);

    std::vector<std::vector<float>> newCamIntrinsics;
    for (int i = 0; i < 3; i++)
    {
        std::vector<float> row;
        for (int j = 0; j < 3; j++)
        {
            row.push_back(newMatrix.at<double>(i, j));
        }
        newCamIntrinsics.push_back(row);
    }
    std::vector<float> newDistCoeffs;
    newDistCoeffs.resize(8);
    std::fill(newDistCoeffs.begin(), newDistCoeffs.end(), 0);
    sensor_msgs::CameraInfo newCameraData = createCameraInfo(_width, _height, newCamIntrinsics, newDistCoeffs);

    ros::Time rosBaseTime = ros::Time::now();
    std::chrono::time_point<std::chrono::steady_clock> steadyBaseTime = std::chrono::steady_clock::now();
    updateBaseTime(steadyBaseTime, rosBaseTime);

    while (ros::ok())
    {
        auto videoIn = video->get<dai::ImgFrame>();
        newCameraData.header.stamp = getFrameTime(rosBaseTime, steadyBaseTime, videoIn->getTimestamp());
        cameraData.header.stamp = newCameraData.header.stamp;
        cv::Mat frame = videoIn->getCvFrame();
        sensor_msgs::Image img_msg;

        _img_bridge = cv_bridge::CvImage(cameraData.header, sensor_msgs::image_encodings::BGR8, frame);
        _img_bridge.toImageMsg(img_msg);
        _camera_pub.publish(img_msg, cameraData, cameraData.header.stamp);

        // if (_rect_image_pub.getNumSubscribers() > 0)
        // {
        //     cv::Mat frameRect;
        //     cv::remap(frame, frameRect, mapX, mapY, cv::INTER_LINEAR);
        //     _img_bridge = cv_bridge::CvImage(cameraData.header, sensor_msgs::image_encodings::BGR8, frameRect);
        //     _img_bridge.toImageMsg(img_msg);
        //     _rect_image_pub.publish(img_msg);
        // }

        ros::spinOnce();
    }
    return 0;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "depthai_oak_1_w");
    ros::NodeHandle nh;
    DepthAiOak1W node(&nh);
    return node.run();
}

File: ros_ws/bw_tracking_cam/src/bw_tracking_cam/depthai_oak_1_w.py
#!/usr/bin/env python
from __future__ import annotations

import time
from typing import Any, Optional

import depthai as dai
import numpy as np
import rospy
from bw_tools.get_param import get_param
from sensor_msgs.msg import CameraInfo
from std_msgs.msg import Header

from bw_tracking_cam.oak_1_resolution_mode import Oak1ResolutionMode
from bw_tracking_cam.oak_time_helpers import get_frame_time, ros_time_from_nsec
from bw_tracking_cam.tracking_camera_node import ImageSupplier, TrackingCameraNode, config_from_ros_param


class OakImageSupplier(ImageSupplier):
    def __init__(self, camera_name: str, fps: float, exposure_time: int, iso: int, resolution_mode: Oak1ResolutionMode):
        self.camera_name = camera_name
        self.fps = fps
        self.resolution_mode = resolution_mode
        self.exposure_time = exposure_time
        self.iso = iso

        self.camera_info = CameraInfo()
        self.ros_base_time = rospy.Time(0)
        self.steady_base_time = time.perf_counter_ns()
        self.frame_num = 0

    def update_base_time(self, steady_base_time_ns: int) -> None:
        current_ros_time = rospy.Time.now()
        current_steady_time_ns = time.perf_counter_ns()

        # Calculate expected offset in nanoseconds
        expected_offset = current_steady_time_ns - steady_base_time_ns

        # Update ros_base_time
        new_base_time_ns = current_ros_time.to_nsec() - expected_offset
        self.ros_base_time = ros_time_from_nsec(new_base_time_ns)

    def init_pipeline(self, cam_key: dai.CameraBoardSocket) -> dai.Pipeline:
        # Create pipeline
        pipeline = dai.Pipeline()

        # Define source and output
        cam_rgb = pipeline.create(dai.node.ColorCamera)
        xout_video = pipeline.create(dai.node.XLinkOut)

        xout_video.setStreamName("video")

        # Properties
        cam_rgb.setBoardSocket(cam_key)
        cam_rgb.setResolution(self.resolution_mode.to_dai())
        cam_rgb.setFps(self.fps)

        xout_video.input.setBlocking(False)
        xout_video.input.setQueueSize(1)

        # Linking
        cam_rgb.video.link(xout_video.input)

        control_in = pipeline.create(dai.node.XLinkIn)
        control_in.setStreamName("control")
        control_in.out.link(cam_rgb.inputControl)

        return pipeline

    def make_camera_info(self, calib_data: dai.CalibrationHandler, cam_key: dai.CameraBoardSocket) -> CameraInfo:
        instrinsics, info_width, info_height = calib_data.getDefaultIntrinsics(cam_key)

        camera_info = CameraInfo()
        camera_info.header.frame_id = self.camera_name
        camera_info.width = info_width
        camera_info.height = info_height
        instrinsics_array = np.array(instrinsics)
        projection = np.zeros((3, 4))
        camera_info.K = instrinsics_array.reshape(9).tolist()
        camera_info.D = np.array(calib_data.getDistortionCoefficients(cam_key)).tolist()
        projection[:3, :3] = instrinsics_array[:3, :3]
        camera_info.P = projection.reshape(12).tolist()

        return camera_info

    def __enter__(self) -> OakImageSupplier:
        cam_key = dai.CameraBoardSocket.CAM_A

        pipeline = self.init_pipeline(cam_key)

        # Connect to device and start pipeline
        self.device = dai.Device(pipeline)
        # Camera control
        control_queue = self.device.getInputQueue("control")
        ctrl = dai.CameraControl()
        ctrl.setManualExposure(self.exposure_time, self.iso)
        control_queue.send(ctrl)

        self.video = self.device.getOutputQueue(name="video", maxSize=1, blocking=False)

        calib_data = self.device.readCalibration()
        self.camera_info = self.make_camera_info(calib_data, cam_key)

        self.update_base_time(self.steady_base_time)

        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        del self.video
        del self.device

    def get_image(self) -> tuple[Header, Optional[np.ndarray]]:
        video_in = self.video.get()
        frame: np.ndarray = video_in.getCvFrame()  # type: ignore

        frame_time_nsec = int(video_in.getTimestamp().total_seconds() * 1e9)  # type: ignore

        timestamp = get_frame_time(self.ros_base_time, self.steady_base_time, frame_time_nsec)
        header = Header(stamp=timestamp, frame_id=self.camera_name, seq=self.frame_num)

        self.frame_num += 1
        return header, frame

    def get_info(self) -> CameraInfo:
        return self.camera_info


class DepthAiOak1W:
    def __init__(self):
        self.fps = get_param("~fps", 60.0)
        self.exposure_time = get_param("~exposure_time", 8000)  # microseconds
        self.iso = get_param("~iso", 800)
        self.resolution_mode = Oak1ResolutionMode("mode_" + get_param("~resolution_mode", "1080_p"))
        self.config = config_from_ros_param()

    def run(self):
        with OakImageSupplier(
            self.config.camera_name, self.fps, self.exposure_time, self.iso, self.resolution_mode
        ) as video:
            node = TrackingCameraNode(self.config, video)

            while not rospy.is_shutdown():
                node.tick()


if __name__ == "__main__":
    rospy.init_node("depthai_oak_1_w", log_level=rospy.DEBUG)
    DepthAiOak1W().run()
